dist-tracing/src/trace.rs

```diff
@@ -0,0 +1,255 @@
+                    dispatch.downcast_ref::<crate::telemetry_layer::TraceCtxRegistry>()
+                {
+                    layer.record_trace_ctx(self, current_span_id.clone());
+                    Ok(())
+                } else {
+                    Err(TraceCtxError::TelemetryLayerNotRegistered)
+                }
+            })
+            .ok_or(TraceCtxError::NoEnabledSpan)?;
+        res
+    }
+
+    /// get the current trace context (as registered on this node or a parent therof
+    /// via 'record_on_current_span'). Requires that the currently registered dispatcher
+    /// have some 'TelemetryLayer' reachable via 'downcast_ref.
+    pub fn current_trace_ctx() -> Result<Self, TraceCtxError> {
+        fn inner(x: (&tracing::Id, &tracing::Dispatch)) -> Result<TraceCtx, TraceCtxError> {
+            let (current_span_id, dispatch) = x;
+            let trace_ctx_registry = dispatch
+                .downcast_ref::<crate::telemetry_layer::TraceCtxRegistry>()
+                .ok_or(TraceCtxError::TelemetryLayerNotRegistered)?;
+
+            let registry = dispatch
+                .downcast_ref::<tracing_subscriber::Registry>()
+                .ok_or(TraceCtxError::RegistrySubscriberNotRegistered)?;
+
+            let iter = itertools::unfold(Some(current_span_id.clone()), |st| match st {
+                Some(target_id) => {
+                    // TODO: confirm panic is valid here
+                    // failure here indicates a broken parent id span link, panic is valid
+                    let res = registry
+                        .span(target_id)
+                        .expect("span data not found during eval_ctx for current_trace_ctx");
+                    *st = res.parent().map(|x| x.id());
+                    Some(res)
+                }
+                None => None,
+            });
+
+            trace_ctx_registry
+                .eval_ctx(iter)
+                .map(|x| TraceCtx {
+                    parent_span: Some(trace_ctx_registry.span_id(current_span_id.clone())),
+                    trace_id: x.trace_id,
+                })
+                .ok_or(TraceCtxError::NoParentNodeHasTraceCtx)
+        }
+
+        let span = tracing::Span::current();
+        let res = span
+            .with_subscriber(inner)
+            .ok_or(TraceCtxError::NoEnabledSpan)?;
+        res
+    }
+}
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct SpanId {
+    pub tracing_id: tracing::Id,
+    pub instance_id: u64,
+}
+
+// TODO: round trip property test for this
+impl SpanId {
+    /// attempts to parse a SpanId from a string
+    pub fn from_string(s: &str) -> Option<SpanId> {
+        let mut iter = s.split('-');
+        let s1 = iter.next()?;
+        let u1 = u64::from_str_radix(s1, 10).ok()?;
+        let s2 = iter.next()?;
+        let u2 = u64::from_str_radix(s2, 10).ok()?;
+
+        Some(SpanId {
+            tracing_id: tracing::Id::from_u64(u1),
+            instance_id: u2,
+        })
+    }
```

Comment:
nit, take it or leave it: if this was in an implementation of the `FromStr` trait for `SpanId`, you could say
```rust
"some string".parse::<SpanId>()
```
which is kinda nice